agatha/construct/semrep_util.py

```diff
@@ -306,6 +306,40 @@ def _get_flags(self, input_path:Path, output_path:Path)->List[str]:
     res.append(str(output_path))
     return res
 
+  def _run_gracefully(self, input_path:Path, output_path:Path)->None:
+    """
+    Runs semrep safely. If a sentence causes a semrep error, 
+    logs the error and re-tries execution with next input.
+    Overall, encapsulate run functionality and use @320 as a
+    wrapper.
+    Args:
+      input_path: Path to the Semrep input file
+    """
+    if not self.metamap_server.running():
+      self.metamap_server.start()
+
+    input_path = Path(input_path)
+    assert input_path.is_file(), f"Failed to find {input_path}"
+    assert not output_path.exists(), f"Refusing to overwrite {output_path}"
+    cmd = self._get_flags(input_path, output_path)
+    env = self._get_env()
+    print("Running:", " ".join(cmd))
+
+    semrep_proc = subprocess.Popen(
+        cmd,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE)
+    _, error = semrep_proc.communicate()
+    if semrep_proc.returncode != 0:
+      # log to file maybe?
+      print('Exception encountered in child process, attempting to respawn')
+      print('Error:{}'.format(error))
+      # recursive call for the next input?
+      self._run_gracefully(input_path, output_path)
```

Comment:
I like the idea, but right now this is an infinite loop. `_run_gracefully` should probably accept a list of strings instead of a file path.  You can modify `run` in order to create that list of strings from the `input_path`.

Maybe instead of a recursive call within `_run_gracefully` we could just throw an exception that `run` catches. Then, `run` can switch to a recovery mode or something?

Reply:
So just pass in the individual lines of the input file and see if that crashes semrep
If it does, we'll raise an exception and catch it in `run()`